machine/src/chip.rs

```diff
@@ -22,31 +19,310 @@ pub trait Chip<M: Machine> {
         vec![]
     }
 
-    fn global_sends(&self, machine: &M) -> Vec<Interaction<M::F>> {
+    fn global_sends(&self, _machine: &M) -> Vec<Interaction<M::F>> {
         vec![]
     }
 
-    fn global_receives(&self, machine: &M) -> Vec<Interaction<M::F>> {
+    fn global_receives(&self, _machine: &M) -> Vec<Interaction<M::F>> {
         vec![]
     }
+
+    fn all_interactions(&self, machine: &M) -> Vec<(Interaction<M::F>, InteractionType)> {
+        let mut interactions: Vec<(Interaction<M::F>, InteractionType)> = vec![];
+        interactions.extend(
+            self.local_sends()
+                .into_iter()
+                .map(|i| (i, InteractionType::LocalSend)),
+        );
+        interactions.extend(
+            self.local_receives()
+                .into_iter()
+                .map(|i| (i, InteractionType::LocalReceive)),
+        );
+        interactions.extend(
+            self.global_sends(machine)
+                .into_iter()
+                .map(|i| (i, InteractionType::GlobalSend)),
+        );
+        interactions.extend(
+            self.global_receives(machine)
+                .into_iter()
+                .map(|i| (i, InteractionType::GlobalReceive)),
+        );
+        interactions
+    }
+
+    fn interaction_map(&self, machine: &M) -> BTreeMap<BusArgument, Vec<usize>> {
+        let mut map: BTreeMap<BusArgument, Vec<usize>> = BTreeMap::new();
+        for (n, (interaction, _)) in self.all_interactions(machine).iter().enumerate() {
+            map.entry(interaction.argument_index)
+                .or_insert_with(Vec::new)
+                .push(n);
+        }
+        map
+    }
+}
+
+pub trait ValidaAir<AB: PermutationAirBuilder, M: Machine> {
+    fn eval(&self, builder: &mut AB, machine: &M);
+}
+
+pub trait ValidaAirBuilder: PermutationAirBuilder {
+    type PublicInput;
```

Comment:
I don't think we'll have other public inputs, what do you think of calling it `fn final_sums(&self)` and having it return something concrete like `&[Self::FE]`?

Reply:
I think you're right, but not sure how I feel about returning a slice, and having to keep track of which sum belongs to which chip. Maybe it's not a big deal.